cpp/src/parquet/column_reader.h

```diff
@@ -302,8 +303,274 @@ class TypedColumnReader : public ColumnReader {
+      cnt += range.count();
+    }
+    return cnt;
+  }
+
+  bool isValid() const {
+    if (ranges.size() == 0) {
+      return false;
+    }
+    if (ranges[0].from < 0) {
+      return false;
+    }
+    for (size_t i = 1; i < ranges.size(); i++) {
+      if (ranges[i].from <= ranges[i - 1].to) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool isOverlapping(int64_t from, int64_t to) const {
+    const Range searchRange(from, to);
+    return isOverlapping(searchRange);
+  }
+
+  bool isOverlapping(const Range& searchRange) const {
+    auto it = std::lower_bound(
+        ranges.begin(), ranges.end(), searchRange,
+        [](const Range& r1, const Range& r2) { return r1.isBefore(r2); });
+    return it != ranges.end() && !(*it).isAfter(searchRange);
+  }
+
+  std::vector<Range>& getRanges() { return ranges; }
+
+  const Range& operator[](size_t index) const { return ranges[index]; }
+
+  std::string toString() const {
+    std::string result = "[";
+    for (const Range& range : ranges) {
+      result +=
+          "(" + std::to_string(range.from) + ", " + std::to_string(range.to) + "), ";
+    }
+    if (!ranges.empty()) {
+      result = result.substr(0, result.size() - 2);
+    }
+    result += "]";
+    return result;
+  }
+};
+
+using RowRangesPtr = std::shared_ptr<RowRanges>;
+
 namespace internal {
 
+class PARQUET_EXPORT RecordSkipper {
+ public:
+  RecordSkipper(RowRanges& pages, RowRanges& row_ranges_)
+      : row_ranges(row_ranges_) {  // copy row_ranges
+    RowRanges will_process_pages, skip_pages;
+    for (auto& page : pages.getRanges()) {
+      if (!row_ranges.isOverlapping(page)) {
+        skip_pages.add(page, false);
+      }
+    }
+
+    /// Since the skipped pages will be slienly skipped without updating
+    /// current_rg_processed_records or records_read_, we need to pre-process the row
+    /// ranges as if these skipped pages never existed
+    adjust_ranges(skip_pages, row_ranges);
+
+    total_rows_to_process = pages.rowCount() - skip_pages.rowCount();
+  }
+
+  /// \brief Return the number of records to read or to skip
+  /// if return values is positive, it means to read N records
+  /// if return values is negative, it means to skip N records
+  /// if return values is 0, it means end of RG
+  int64_t advise_next(const int64_t current_rg_procesed) {
+    if (row_ranges.getRanges().size() == row_range_idx) {
+      return 0;
+    }
+
+    if (row_ranges[row_range_idx].to < current_rg_procesed) {
+      row_range_idx++;
+      if (row_ranges.getRanges().size() == row_range_idx) {
+        // negative, skip the ramaining rows
+        return current_rg_procesed - total_rows_to_process;
+      }
+    }
+
+    if (row_ranges[row_range_idx].from > current_rg_procesed) {
+      // negative, skip
+      return current_rg_procesed - row_ranges[row_range_idx].from;
+    }
+
+    const auto ret = row_ranges[row_range_idx].to - current_rg_procesed + 1;
+    assert(ret > 0);
```

Comment:
assert generally aren't used in the code base.  use either DHECK or raise an exception.

Reply:
done